qemu/tests/watchdog.py

```diff
@@ -282,6 +283,51 @@ def hotplug_unplug_watchdog_device():
                 test.fail("Oops, find watchdog pci, unplug failed")
             logging.info("The WDT remove successfully")
 
+    def heartbeat_test():
+        """
+        Heartbeat test for i6300esb
+        Test steps:
+        1.Start VM with "-watchdog-action pause" CLI option
+        2.Set heartbeat value and reload the i6300esb module
+        3.Trigger wathchdog action through open /dev/watchdog
+        4.Ensure watchdog_action takes effect after $heartbeat.
+        """
+
+        del_module_cmd = params["del_module_cmd"]
+        reload_module_cmd = params["reload_module_cmd"]
+        _watchdog_device_check(test, session, watchdog_device_type)
+
+        error_context.context("set heartbeat value and reload the i6300esb "
+                              "module", logging.info)
+        session.cmd(del_module_cmd)
+        heartbeat = random.randint(1, 180)
+        session.cmd_output(reload_module_cmd % heartbeat)
+        _watchdog_device_check(test, session, watchdog_device_type)
+
+        _trigger_watchdog(session, trigger_cmd)
+
+        error_context.context("Judge watchdog fire time", logging.info)
+        start_time = time.time()
+        end_time = start_time + float(heartbeat)
+        while not vm.monitor.verify_status("paused"):
+            if time.time() > end_time + 2:
```

Comment:
Could the "later than expected" fail been raised here directly? We have same judgement in line 327 anyway, we don't have to go through it again. Like:
```
if time.time() > end_time + 2:
    test.fail("... later than expected.")
```

Reply:
They are different. Line 313 represents watchdog doesn't take effect after $heartbeat, while Line 327 represents watchdog takes effect after $heartbeat.

Reply:
@YanhuiMa Then it should equal to the judgement in line 317, here we can fail it with "watchdog action didn't take effect"